### PR TITLE
Implementing two-phase grain growth model

### LINKED ISSUE #8837: generalizing the single-phase grain growth model to two-phase systems
### Description of the enhancement or error report

Extending the single-phase grain growth models to two-phase systems, where each phase can have multiple orientations based on Moelans model (Moelans Acta Mat 2011). 

### Rationale for the enhancement or information for reproducing the error
It can be used for investigating grain growth in two-phase systems.  This is also an important step for multi-phase phase-field models, where multiple species concentrations can also be coupled to that model.  

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

### PR DESCRIPTION
**Pull request by karim-E-ahmed**
Implementing a two-phase grain growth model.

closes #8837

### CODE REVIEW

**Comment by dschwen** [comment]
I'd prefer it if you derived a class from `PolycrystalKernelAction` (for example `PolycrystalTwoPhaseKernelAction`).

File: modules/phase_field/src/action/PolycrystalKernelAction.C [hunk 1]
```diff
@@ -67,6 +73,13 @@ PolycrystalKernelAction::act()
       params.set<bool>("use_displaced_mesh") = getParam<bool>("use_displaced_mesh");
       if (isParamValid("T"))
         params.set<std::vector<VariableName>>("T") = {getParam<VariableName>("T")};
+      params.set<unsigned int>("op_index") = op;
+      params.set<unsigned int>("second_phase_op_num") =
+          getParam<unsigned int>("second_phase_op_num");
+      params.set<Real>("en_ratio") = getParam<Real>("en_ratio");
```
Comment (dschwen) [inline modules/phase_field/src/action/PolycrystalKernelAction.C]: replace all this with `applyParameters(parameters());`

**Comment by dschwen** [comment]
@karim86 did you plan on fixing this up?

**Comment by karim86** [comment]
Yes, I should be able to finish it next week.

On Fri, Apr 14, 2017 at 11:40 AM, Daniel Schwen <notifications@github.com>
wrote:

> @karim86
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_karim86&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=nhScyjGW-UG7fBfM7LNyTofBdOEwu-Evh358cPutz28&m=hQ4deLNFbHrF84qpT_0qIDvjFgAcFsyZie2hwPtLJXE&s=fMtgHDrX-Onn6IwkgEMWiZG1xo5Nk5_kv111vf-FJcw&e=>
> did you plan on fixing this up?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_idaholab_moose_pull_8838-23issuecomment-2D294199257&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=nhScyjGW-UG7fBfM7LNyTofBdOEwu-Evh358cPutz28&m=hQ4deLNFbHrF84qpT_0qIDvjFgAcFsyZie2hwPtLJXE&s=zdT7PiIuqAk6F23nwV-y7Lm_j1k5RfXUCijernLjjCw&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AOW-5F8-2DG2fBvX5AQMV6ipIt6lxevIfHVxks5rv6-2DagaJpZM4MtiZv&d=DwMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=nhScyjGW-UG7fBfM7LNyTofBdOEwu-Evh358cPutz28&m=hQ4deLNFbHrF84qpT_0qIDvjFgAcFsyZie2hwPtLJXE&s=pprDGo1JkhrAr-BSz6NB_3tcbem66blsSK1_szkVn0Q&e=>
> .
>



-- 
Karim El-Sayed Ahmed, PhD
Postdoctoral Research Associate
Fuel Modeling & Simulation Department
Idaho National Laboratory

**Comment by karim86** [comment]
@dschwen How to default the get kernel type () in the original polycrestal action so that if we do not use polycrestal two-phase action it will garb the original ACGrGrPoly (I made TowPhase ACGrGrPoly for the two-phase case)

**Comment by karim86** [comment]
@dschwen I'm just gonna use phase_num in the original action, so if its one then it looks for ACGrGrPoly otherwise it looks for the corresponding kernel for two or more phases.

**Pull request closed without merging**